iot-hub/Samples/service/IotHubSasCredentialAuthenticationSample/Program.cs

```diff
@@ -0,0 +1,87 @@
+                })
+                .WithNotParsed(errors =>
+                {
+                    Environment.Exit(1);
+                });
+
+            // Initialize SAS token credentials.
+            Console.WriteLine("Creating sas credential.");
+
+            TimeSpan timeToLive = TimeSpan.FromHours(1);
+            DateTime expiresOn = DateTime.UtcNow.Add(timeToLive);
+            string sasToken = GenerateSasToken(parameters.HostName, parameters.SharedAccessKey, expiresOn);
+            // Note: Pass the generated sasToken and not just the shared access signature when creating the AzureSasCredential.
+            AzureSasCredential sasCredential = new AzureSasCredential(sasToken);
+
+            // This is how the credential can be updated in the IotHubSasCredential object whenever necessary.
+            // This sample just shows how to perform the update but it is not necessary to update the token
+            // until the token is close to its expiry.
+            // NOTE: Ensure to pass the new expiry date time.
+            DateTime newExpiresOn = DateTime.UtcNow.Add(timeToLive);
+            string updatedSasToken = GenerateSasToken(parameters.HostName, parameters.SharedAccessKey, expiresOn);
+            sasCredential.Update(updatedSasToken);
+
+            // There are constructors for all the other clients where you can pass SAS credentials - JobClient, RegistryManager, DigitalTwinClient
+            using var serviceClient = ServiceClient.Create(parameters.HostName, sasCredential, parameters.TransportType);
+
+            var sample = new IotHubSasCredentialAuthenticationSample();
+            await sample.RunSampleAsync(serviceClient, parameters.DeviceId);
+        }
+
+        private static string GenerateSasToken(string hostName, string sharedAccessKey, DateTime expiresOn, string policyName = default )
+        {
+            DateTime epochTime = new DateTime(1970, 1, 1);
+            TimeSpan secondsFromEpochTime = expiresOn.Subtract(epochTime);
+            long seconds = Convert.ToInt64(secondsFromEpochTime.TotalSeconds, CultureInfo.InvariantCulture);
+            string expiry = Convert.ToString(seconds, CultureInfo.InvariantCulture);
+
+            string stringToSign = WebUtility.UrlEncode(hostName) + "\n" + expiry;
+
+            HMACSHA256 hmac = new HMACSHA256(Convert.FromBase64String(sharedAccessKey));
+            string signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(stringToSign)));
+
+            // SharedAccessSignature sr=ENCODED(dh://myiothub.azure-devices.net/a/b/c?myvalue1=a)&sig=<Signature>&se=<ExpiresOnValue>[&skn=<KeyName>]
+            string token = string.Format(
+                CultureInfo.InvariantCulture,
+                "SharedAccessSignature sr={0}&sig={1}&se={2}",
+                WebUtility.UrlEncode(hostName),
+                WebUtility.UrlEncode(signature),
+                expiry);
+
+            if (!string.IsNullOrWhiteSpace(policyName))
```

Comment:
Good idea making the policy name an optional parameter. I need to do the same on the Java sample code.